src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Unix.cs

```diff
@@ -150,6 +150,7 @@ internal static unsafe IntPtr AllocBSTR(int length)
 
             IntPtr s = p + sizeof(IntPtr);
             *(((uint*)s) - 1) = (uint)(length * sizeof(char));
+            ((char*)s)[length] = '\0';
```

Comment:
In C++, `AllocBSTR` forwards the arguments to `AllocBSTRByteLen` https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/palrt/bstr.cpp#L123 which does `NULL-terminate with both a narrow and wide zero.` as we have below on line 178. Should we do the same here or replace two usages of AllocBSTR in `SecureString.cs` and `Marshal.cs` with `AllocBSTRByteLen`?

Reply:
> NULL-terminate with both a narrow and wide zero

This only makes difference for legacy BSTRs with odd byte at the end. The regular BSTRs that this path allocates do not need this - both narrow and wide zero is in the same place in memory for them, so it is enough to just write the wide zero here.

Reply:
Thanks for the clarification. would it make sense to update AllocBSTRByteLen with the same approach to align the two implementations and deduplicate these methods? ByteLen variant was just a rewrite of C++ version, where both Alloc* methods do the same thing.

Reply:
As far as I can tell, the C++ version of these methods is factored in the same way as this managed version:
- https://github.com/dotnet/runtime/blob/master/src/coreclr/src/palrt/bstr.cpp#L151
- https://github.com/dotnet/runtime/blob/master/src/coreclr/src/palrt/bstr.cpp#L100

I think this small duplication is ok.